### ISSUE TITLE
Subscribe to a toot

### ISSUE THREAD

**Issue by allo-** [tootsuite/mastodon#7969]
An option to subscribe to a toot would be useful, where you can subscribe to notifications about the toot.

Example toot: "Hey, does somebody know a command line tool to do X?"

I would now like to be able to subscribe to the notifications, to see when the original poster gets an useful answer which is interesting for me as well.

It may be interesting to get all notifications (like boosts and favs), but I do not see many usecases for that. The most interesting part are notifications about further replies.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by Elshara** [comment]
I second this. People should be able to control this in notification settings.

**Comment by cbayerlein** [comment]
I second!

**Comment by jpfox** [comment]
Yes it's a very good feature !

**Comment by allo-** [comment]
bump

**Comment by rugk** [comment]
I like this idea.

**Comment by resynth1943** [comment]
I don't think Twitter can do this. I love the idea!

**Comment by placoderm** [comment]
I would also like this.

I'm new using mastodon, so I don't understand how I am supposed to stay engaged in a new post that comes into my home thread. Obviously there won't be any comments as soon as it pops up. But what if I want to see later what other people have written? Do I just have to bookmark it and then remember to go back and look?

**Comment by privacy1st-de** [comment]
I've written a Python program that generates an RSS feed with news from a Toot (new replies and edits). Maybe this is helpful in the meantime:  https://mastodon-toot-follower.privacy1st.de/ 

---

I was able to learn some Flask and Jinja2 along the way ... so please excuse that I have not directly contributed Mastodon 😅

**Comment by nd-** [comment]
Very good idea, I would use it very often, for example when someone asks a question and I'd like to see the answers.

**Comment by turvanen** [comment]
I'd like to have this but also sometimes to get notifications only when the thread is being expanded (ie. exclude replies by other users). It'd be very useful for following updating news threads, long link collections, follow recommendations etc, especially in cases when they are updated only rarely.

**Comment by brendanjones** [comment]
So that we're clear, what are people expecting when they follow a toot?

Would you want to get notified about:

1. Top-level replies only, by the author only
2. Top level replies only, plus replies by the author at any depth
3. Top level replies only
4. All replies at any depth

**Comment by placoderm** [comment]
I'd be happy with #4. Either I'm interested in the conversation or not.

**Comment by nd-** [comment]
Same here, 4 (all replies) seems more logic and usefull to me.

**Comment by robsonsobral** [comment]
I would be happy with the option 4. Sure, the conversation can go far beyond what I'm interested in, but I can just stop to follow it.

**Comment by yoshimo** [comment]
If ALL is available, the client can still filter down and add context around it if need be.

**Comment by kefir-** [comment]
As the author of duplicate issue #25503 and contributor to duplicate issue #23702, I think number 4 is the feature that is worth pursuing. I often see interesting questions in my feed, and I want to watch how the discussion develops but there's currently no way to "subscribe" for notifications. Both Facebook and friendica have this feature already. 

Manual workarounds like keeping an open tab in the browser for each interesting thread, or adding each post as a favourite and checking back regularly for new replies aren't realistic ways to build a community.

**Comment by Mart-Bogdan** [comment]
I think ideally all 4 would be useful.

And let user chose. Filtering could be done on user's server or at client side. 

Would be really nice to have

**Comment by Karmalakas** [comment]
Wonder if there's any chance of having that

**Comment by kefir-** [comment]
Doesn't seem likely, at this point? My own thinking is that this would be very useful and engaging functionality, and it would deserve a fairly prominent button, probably should replace "save" next to "boost" and "like". I would be surprised if a user experiment shows that "save" is more used than "subscribe to thread" over time. 

Interestingly, even github itself has a button to subscribe to threads. It's a critical core feature, otherwise you lose engagement. Seems mastodon devs don't agree, and I don't have the opportunity to contribute myself. 🤷‍♂️

**Comment by shaedrich** [comment]
I find "save" hella confusing in the context of subscribing. I don't care if it's called "subscribe", "notify me", "follow" or whatever, but "save" sounds like bookmarking and therefore would have nothing to do with notifications.

**Comment by kefir-** [comment]
Perhaps I've misunderstood your point, but I didn't intend for "Save" to mean that notifications should be sent. I personally don't use the "save" button much, but I would use a "follow" a lot.

By the way, someone mentioned "boost and follow". I think this could be two separate buttons, and I think I'd tend to click both or only "follow". If I boost something, and the discussion sheds new light on the post and maybe even shows that the original was wrong, I'd like to know, and then maybe I can also share the updated information.

**Comment by shaedrich** [comment]
Okay, maybe, there's a bit of miscommunication here. Do you mean "bookmark" by "save"? Because in the web interface, the 🔖 emoji is called "bookmark". People debate about it's "proper" implementation but that's a topic for another day. However, I use it a lot (more than I should, actually 😅).

Currently, both boost and fav result in a follow. This doesn't necessarily need to change. They would become "implicit follow", while the 🔔 emoji (or whatever would be used), would become "explicit follow". Well, the more expectable label would probably be "follow only".

**Comment by kefir-** [comment]
Yes, when I said "save" I guess I mean "bookmark", which I don't use much. 

But you say boost and fav result in a follow? What follow is that? I don't see any notifications to replies after boosting or liking a post.

**Comment by Karmalakas** [comment]
> Currently, both boost and fav result in a follow

Are you sure? Neither boost nor fav (or save/bookmark for that matter) does not end up in notifications if somebody replies. The only way to get notifications that I know of, is to comment, and even then you'll be notified only is someone replies to your comment

**Comment by shaedrich** [comment]
Yeah, okay, maybe, there's another issue here: What _is_ a subscription? There could be different types/levels of notifications, similar to the toot visibility level.

You don't get notified about new replies, however, you do, when the toot is updated. So, there's some kind of subscription ("subscription light" if you will). But that might not be what we want here.

Long story short: At some point, we might need to have precise definitions of terms, we all agree on.

**Comment by kefir-** [comment]
Yes, I've seen notifications about updated toots.

But this issue, and several others that have been closed as duplicates, is about getting notifications when an interesting post/question gets a response. Just like the notification emails you're getting from github as I post this response. Otherwise none of us would probably ever check back here, and this discussion would have been dead a long time ago.

**Comment by brendanjones** [comment]
> Long story short: At some point, we might need to have precise definitions of terms, we all agree on.

Yes indeed. Here we go:

We already get notifications about edited posts that we've boosted (and fav'd? Can't remember off the top of my head), so I think the various functions that can be added here are:

1. Subscribe to post edits (independent of having to boost/fav the post)
2. Subscribe to top-level replies by post author only
3. Subscribe to all replies by post author only
4. Subscribe to all top-level replies
5. Subscribe to all replies

I think that's all (sensible) variations - if anyone has any more then please suggest and I'll edit them in.

**Comment by brendanjones** [comment]
A comment on "1. Subscribe to post edits (independent of having to boost/fav the post)": I can see that this might be useful in some circumstances but I'd argue against implementing it for two reasons:

1. I'm not sure we should be encouraging post edits. It's good there is an edit function, but do you want to have to check edit notifications to see whether the OP was just correcting an error or actually adding new information? Nah, keep post edits for correcting errors only.
2. Simplicity of user actions. Assuming (2) or (3) get implemented, that makes two ways for people to update post threads with new information. Which one do you choose when you want to be updated on a post? I think better to have only one way of updating posts.

**Comment by kefir-** [comment]
I'll argue for only nr 5. We're deep in a discussion here, for the average user to understand all these differences, this requires some serious UI design and education. That's in addition to all the maintenance by devs. Just subscribe, and then unsubscribe when you don't want to follow any more. I think this is what other platforms do too, like github thread subscription, Facebook post notifications, etc. It's the least surprising functionality for users, IMO.

That being said, I think post edits might warrant notifications too in this case, so you know if people are responding to a different text than you read.

**Comment by shaedrich** [comment]
Similar to how you select the post visibility
![grafik](https://github.com/user-attachments/assets/9fd208d4-a373-46a9-b517-8da4be029ef8)
one could select the subscription level
![grafik](https://github.com/user-attachments/assets/8b02e27f-2cca-413a-b9f6-0ccf2b8fe40e)

**Comment by kefir-** [comment]
Not bad!

But even if all those are possible in theory, do they all have probable use cases? I still think the most likely use case is the last option, if you wanted to follow a discussion.

**Comment by robsonsobral** [comment]
Isn't better, for users and devs, to start simple?

**Comment by nd-** [comment]
> I'll argue for only nr 5. We're deep in a discussion here, for the average user to understand all these differences, this requires some serious UI design and education. That's in addition to all the maintenance by devs.

+1

**Comment by lgs11** [comment]
Here's a big tangent. Forty years ago, we were using the text and keyboard driven Usenet readers like trn (threaded news reader), which worked in the opposite way of today's social media and forum apps.

One message at a time would show, and the user had to choose between two different commands to continue, either 'n' for next, meaning keep showing new messages in this thread, or 'k' for kill, meaning don't show messages from this thread again.

This was very efficient and much faster for the user than scrolling and clicking buttons. Of course it wouldn't be very useful for touch screens, but as an alternative way to navigate in Mastodon by keyboard, it could prove very productive. Maybe as a selectable mode in the current viewer.  There would be keyboard commands like 'c' that would catch up on all unread replies, 'r' to reply, 'R' to reply to all, and others to mute or block users, etc.

It would take a bit of time to learn, but it's not hard to implement. The main difference is to always show replies to messages I have read, unless I "kill" it.

https://en.wikipedia.org/wiki/Rn_(newsreader)

**Comment by shaedrich** [comment]
@lgs11 Interesting. However, I think, this would be more fitting as its own frontend. And I don't think, this issue is the proper place for it.

**Comment by trwnh** [comment]
@brendanjones wrote:

> We already get notifications about edited posts that we've boosted (and fav'd? Can't remember off the top of my head), so I think the various functions that can be added here are:
> 
>     1. Subscribe to post edits (independent of having to boost/fav the post)
> 
>     2. Subscribe to top-level replies by post author only
> 
>     3. Subscribe to all replies by post author only
> 
>     4. Subscribe to all top-level replies
> 
>     5. Subscribe to all replies
> 
> 
> I think that's all (sensible) variations - if anyone has any more then please suggest and I'll edit them in. I do not know the technical feasibility of implementing these, I'm only listing possible user actions.

From a technical feasibility standpoint, you would need something concrete to follow explicitly. This means turning the object into an actor. I'm not sure it makes sense to follow a post, since that's not really semantically clear what it entails -- it would be up to the post itself (or by proxy the post author) to decide what to put in the `outbox` of that post and what to send to its `followers`. Perhaps the post could `Announce` or inbox-forward any activity that it receives or is relevant to it?

A more restricted subset of that, which makes more sense and is more theoretically developed, would be to follow a Collection specifically. The Collection owner could generate Add/Remove activities targeting that collection (and if the Collection owns itself, then it could generate those activities itself).

Given that, you could then follow any of the below collections if they exist and are specified (and are followable actors):

- `likes` (as specified in ActivityPub, this contains Like activities where the `object` is the current object)
- `shares` (as specified in ActivityPub, this contains Announce activities where the `object` is the current object)
- `replies` (as specified in [FEP-7458](https://w3id.org/fep/7458), this contains any object where `inReplyTo` is the current object)
- `context` (as specified in [FEP-7888](https://w3id.org/fep/7888), this contains any object that has been added to the conversation/thread/topic)

To be able to follow any particular subset of activity surrounding a post, you would need to have an existing collection that supports being followed. Collections that don't exist can instead be defined as an extension property:

- [FEP-bad1](https://w3id.org/fep/bad1) defines a `history` collection (name pending) that contains all activities where the `object` is the current object. This would allow viewing the original Create, any Updates, as well as potentially any Likes and Announces etc etc. The only thing it would *not* contain are replies, since those are Create activities with a different `object`.
- You could similarly define a collection only for `revisions`, which contains the Create activity and any Updates... or it could contain something else entirely, like an `ObjectRevision` object that represents the object at that given point in time. (I think Litepub had a `formerRepresentations` collection which worked vaguely similarly to this.)
- ...

**Comment by shaedrich** [comment]
Sorry for my stupidity but I'm afraid, I only got half of it if even that much

**Comment by rubcap** [comment]
+1
Being able to _subscribe_ to a toot and get _notified_ when someone makes a public/unlisted reply, whether or not 
- you follow the OP or the replier(s), 
- you replied or not 

would be very handy!

**Comment by dreua** [comment]
Skimming through all the comments here I saw a lot of discussion on what exactly following would mean and how the UI could look like. I think for a first iteration a simple follow "all replies at any depth" option would be quite useful already combined with an easily accessible way to unsubscribe if it gets too much. Maybe notifications could be compacted to one (like "there are new toots in this thread you are following" -> link to thread or latest toot therein) instead of one notification per toot.

But I haven't seen much from the technical side: How could it be implemented? Does it need a change in protocol? Is it feasible at all performance wise?

**Comment by allo-** [comment]
That's a good point, because the events need to arrive at the subscriber's server.

When the author edits a toot, people who faved or boosted it already get a notification about it. So when the author's server gets a notification about a reply, it would need to forward to the subscribers it in a similar fashion.

**Comment by miblodelcarpio** [comment]
I've been imagining the implementation of this working somehow like a group / organisation account – containing multiple user accounts able to manage the group's single feed – where only the author has write access. If there's already support for grouping in the protocol, or any work put into that idea, how about that being a starting point for our toot subscription feature?

**Comment by kefir-** [comment]
> But I haven't seen much from the technical side: How could it be implemented? Does it need a change in protocol? Is it feasible at all performance wise?

I think the protocol is sufficient as it is, at least for some basic functionality. There's already a generic concept of following an object:

https://www.w3.org/TR/activitystreams-vocabulary/#dfn-follow

And messages (can) have an indicator of which message(s) they are in reply to:

https://www.w3.org/TR/activitystreams-vocabulary/#dfn-inreplyto

So that could be used to resolve threads if someone is following a topic.

Performance will surely be an issue if the server needs to resolve an arbitrary number of parent posts (for all posts made?) to resolve a thread and check if any have been followed. But inReplyTo can contain multiple references, so servers could look for follows only among the inReplyTo objects, and simply expect the original thread starter post and the direct parent to be included as a minimum for this to be useful. Other intermediate objects could be optional. Then it should be a matter of sending notifications when one of the inReplyTo objects match any registered object follows.
